### ISSUE TITLE
"unless-stopped" containers are not started after daemon restart

### ISSUE THREAD

**Issue by KamilKopaczyk** [moby/moby#36496]
**Description**

Containers with restart policy "unless-stopped" are stopped after restarting docker daemon. Please take a look at `docker info` output, as it's printing confusing information versus what `docker ps` shows.

**Steps to reproduce the issue:**
1. Run container with restart policy set to "unless-stopped"
2. Execute `service docker restart`
3. Execute `docker ps`

**Describe the results you received:**
```
CONTAINER ID        IMAGE                                                                                        COMMAND                  CREATED             STATUS                          PORTS                           NAMES
```

**Describe the results you expected:**
```
CONTAINER ID        IMAGE                                                                                        COMMAND                  CREATED             STATUS                          PORTS                           NAMES
<list of containers that were running before restart>
```

**Additional information you deem important (e.g. issue happens only occasionally):**

--

**Output of `docker version`:**

```
Client:
 Version:	18.02.0-ce
 API version:	1.36
 Go version:	go1.9.3
 Git commit:	fc4de44
 Built:	Wed Feb  7 21:16:37 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.02.0-ce
  API version:	1.36 (minimum version 1.12)
  Go version:	go1.9.3
  Git commit:	fc4de44
  Built:	Wed Feb  7 21:15:09 2018
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**

**Please note that `docker info` shows that 99 containers are running**

```
Containers: 124
 Running: 99
 Paused: 0
 Stopped: 25
Images: 174
Server Version: 18.02.0-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: false
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: 949e6fa
Kernel Version: 4.9.0-0.bpo.5-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 11.85GiB
Name: int
ID: QNEP:YJM4:DFBS:Y7RO:GVQS:YDSB:S77P:SEXL:NZ6X:LI6P:3JFG:4Y6Q
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=generic
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: overlay2: the backing xfs filesystem is formatted without d_type support, which leads to incorrect behavior.
         Reformat the filesystem with ftype=1 to enable d_type support.
         Running without d_type support will not be supported in future releases.
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Linux int 4.9.0-0.bpo.5-amd64 #1 SMP Debian 4.9.65-3+deb9u2~bpo8+1 (2017-01-05) x86_64 GNU/Linux

**Comment by thaJeztah** [comment]
Trying to reproduce this, but seems to work for me;

```bash
docker run -d --name foo --restart=unless-stopped nginx:alpine

docker ps

CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
34a19e5b36de        nginx:alpine        "nginx -g 'daemon of…"   5 seconds ago       Up 4 seconds        80/tcp              foo

systemctl restart docker

 docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
34a19e5b36de        nginx:alpine        "nginx -g 'daemon of…"   35 seconds ago      Up 7 seconds        80/tcp              foo
```

Is there anything in the logs that may indicate why the containers were not started? Were they running successfully before the docker service was restarted?

**Comment by KamilKopaczyk** [comment]
I tried the same scenario again and got very similar result. 
Some of the containers are running, but some are not.

After restart, i did `docker inspect` on one of those containers that were running before daemon restart, here's a relevant fragment of it:
```
"State": {
            "Status": "exited",
            "Running": false,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 255,
            "Error": "container \"b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470\": already exists",
            "StartedAt": "2018-03-06T14:08:26.967772641Z",
            "FinishedAt": "2018-03-06T15:31:54.630338501+01:00"
        },
        "Image": "sha256:8a8674f6632fceaa3e69a3f296895d99fed8ef7773fff4a8c5c33e7b152f059d",
        "ResolvConfPath": "/var/lib/docker/containers/b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470/hostname",
        "HostsPath": "/var/lib/docker/containers/b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470/hosts",
        "LogPath": "/var/lib/docker/containers/b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470/b7444d2773aeff77b7c2dd6cf9492b8bd4538bdd0f01e796c97540b6ee0fd470-json.log",
        "Name": "/<somename>",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "<ournetwork>",
            "PortBindings": {},
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
```

Since i'm able to reproduce it, is there anything i could search for, to track the source of this issue down?

**Comment by thaJeztah** [comment]
Ah, thanks for the additional information. The error `"container ...: already exists"` is the culprit here, which makes this a duplicate of https://github.com/moby/moby/issues/36145

That issue is fixed on "master" through https://github.com/moby/moby/pull/36249, and the fix is included in the upcoming Docker 18.03 release; https://github.com/docker/docker-ce/commit/5d3102854bb63a1a2b56bcdc4ba880c2f280d54d (release candidates for 18.03 are currently available in the "test" channel)


I'll close this issue because this is already fixed in the current code on master, but feel free to continue the conversation 👍
